### ISSUE TITLE
It would be nice if indexing into a StencilArray returned a populated stencil

### ISSUE THREAD

**Issue by alex-s-gardner** [rafaqz/Stencils.jl#32]
As of now indexing into a StencelArray returns the value of the parent array:

```julia
julia> r = rand(10,10);
julia> stencil = Window(1);
julia> A = StencilArray(r, stencil);
julia> A[1]
0.10124501404348107
```

wouldn't it be better (more powerful) if indexing into a StencelArray returned a populated stencil?
```
A[1] == Stencils.stencil(A, (1,1)) 

9-element StaticArraysCore.SVector{9, Float64} with indices SOneTo(9):
 0.0
 0.0
 0.0
 0.0
 0.10124501404348107
 0.7814433661466708
 0.0
 0.5830912334108994
 0.9407895258807342
```

**Comment by rafaqz** [comment]
But then you cant use it as a regular array! 😱

Doing weird things with getindex is Bad™️

**Comment by alex-s-gardner** [comment]
Hmmm... not even if you consider a StencilArray as a lazy representation of a n1*n2*... by p array where the parent array has dimensions of n1,n2,... and the stencil has p elements?

**Comment by rafaqz** [comment]
Yeah it would be nice to have broadcasts be over stencils, true... Maybe we can add an array type that does this?

My corncern is personally I always use these at the same time as using them as nomal arrays. For example I will have an array of organism habitat suitability and use it for a cellwise growth model  *and* to drive dispersal with a stencil. Or the same thing with land cover classes and LUCC models. So indeximg has to be normal.

**Comment by alex-s-gardner** [comment]
After playing around with Stencils some more... its truly impressive.. I haven't seen anything this performant... well done.. 

I suppose you'd need to do a lot of re-plumbing if you changed the functionality of `StencilArray`... if it isn't too painful I'd just modify the existing type as it's already intuitive as to what it is. I suspect most users would will construct a `StencilArray` for indexing into the stencils. Would you not still have access to the parent array with `A.parent`?  Doing this would make `StencilArray` highly compositable and very Julian ... if you need the old functionality could you change the name of the current `StencilArray`? If not then I guess something like `LazyStencilArray` could work. 

As for setindex!, I'd leave that as undefined.

**Comment by rafaqz** [comment]
Thanks! Yep, its fast. There is even more SIMD we are missing too... I have boolean grids running 3x faster single threaded than with KernelAbstractions.jl on 6 cores in DymanicGrids.jl (where this all comes from) because I can get the code to unroll and simd down the colums. It will be less impressive for Float64, of course.

The problem is all the use-cases where you want both things. We need to have both options possible.

So we need 2 names that reflect the use cases. Your tapsk is to think if them pcarefpully... LazyArrays are already something different in julia, and dont really suggest this difference

**Comment by alex-s-gardner** [comment]
would you entertain using `StencilArray` for the Lazy representation? It seems this is the most suitable and intuitive name and would be the most general use case. And `A.parent` won't work for indexing into the original array? Not to be daft but it seems un-intuitive to wrap an array in `StencilArray` and be able to index into the parent. `StencilArray` doesn't create a new copy of the parent array so why would the user not just index into the original array or into `A.parent`? 

If that solution is off the table then I'll need to think of a second-best option of a new Type name

**Comment by alex-s-gardner** [comment]
StencilArray = lazy array
ProtoStencilArray = original 

-or- 

StencilArray = original
TemplateArray = lazy array

**Comment by rafaqz** [comment]
Hm not quite clear to me... what is Template? And why Proto?

What is the key difference between them? How do we describe it?

The reason I made a StencilArray object is so the size is in the type for makes faster loops, and so it can have preallocated halo padding to match its stencil radius, but index like a regular array.

`PaddedArray` is possible but already has a package.

`StencilArray` is nice for your use case as its literally an array if stencils. `ArrayOfStencils` is a nice literal name for it too..

**Comment by alex-s-gardner** [comment]
`PaddedArray` is unintuitive as it implies the same array but padded... a StencilArray is something different

`Template` is the terminology used in computer vision for "Stencils"... but I like `ArrayOfStencils` better

`StencilArray` is ideal for my use case 

`ProtoStencilArray` as it is not yet a full `StencilArray`, but rather all of the information needed to construct a `StencilArray` as defined for my use case... hence `Proto`... 

`ArrayOfStencils` is a good name too, the only con I can think of is that it could be ambiguous with `StencilArray` but this might be the best compromise.

**Comment by rafaqz** [comment]
I think using `stencil(A, I)` is fine for now, should do too much work just to get `[ ]` syntax...

**Comment by alex-s-gardner** [comment]
It's your package ... but I'm still convinced that [] is a super power that should be unleashed

**Comment by rafaqz** [comment]
Yeah I was thinking about broadcasting and reductions also being nice... 

It would have to be read only or it gets too confusing. 

And it's cool, but it does add conceptual overhead and code for two array types rather than one.

**Comment by alex-s-gardner** [comment]
given the nature of stencils, i think read only access should be expected

**Comment by rafaqz** [comment]
Ok, if I really have nothing to do one day, or am avoiding something else I really need to do... We can have ArrayOfStencils

**Comment by alex-s-gardner** [comment]
Ha ha... that sounds like an amazing day ;-)

**Comment by rafaqz** [comment]
I just realised the problem with `broadcast` and `ArrayOfStencil` is `mapstencil` uses KernelAbstractions.jl and regular broadcasts are not threaded. So this approach will be a lot slower than `mapstencil` in many cases.

So could hack `copyto!` to actually use KernelAbstractions.jl and make the broadcast fast... which makes me wonder why we don't have a `KernelAbstractionsArray` wrapper in Julia already which would just do this...
